lib/utils/fs.go

```diff
@@ -241,15 +243,67 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 	return fileLock.Unlock, nil
 }
 
+// RemoveAllSecure is similar to `os.RemoveAll` however will delegate to the `RemoveSecure` implementation below
+func RemoveAllSecure(path string) error {
+	// Match os.RemoveAll protections in not permitting removal of "." directories
+	if path == "." || (len(path) >= 2 && path[len(path)-1] == '.' && os.IsPathSeparator(path[len(path)-2])) {
+		return &os.PathError{Op: "RemoveAllSecure", Path: path, Err: syscall.EINVAL} // error type matches os.RemoveAll
+	}
+
+	if info, err := os.Lstat(path); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		} else {
+			return trace.ConvertSystemError(err)
+		}
+	} else if info.IsDir() {
+		var removeErrors []error
+		if files, err := os.ReadDir(path); err != nil {
+			// don't fail fast, allow removal at end to be attempted
+			removeErrors = []error{err}
+		} else {
+			for _, f := range files {
+				if err := RemoveAllSecure(filepath.Join(path, f.Name())); err != nil {
+					removeErrors = append(removeErrors, err)
+				}
+			}
+		}
+		if err := os.Remove(path); err != nil {
+			removeErrors = append(removeErrors, err)
+		}
+		if len(removeErrors) > 0 {
+			if len(removeErrors) == 1 {
+				return trace.ConvertSystemError(removeErrors[0])
+			} else {
+				return fmt.Errorf("multiple errors in directory removal: %v", removeErrors)
+			}
+		}
+		return nil
+	} else { // file or symlink
+		return RemoveSecure(path)
+	}
+}
+
 // RemoveSecure attempts to securely delete the file by first overwriting the file with random data three times
 // followed by calling os.Remove(filePath).
 func RemoveSecure(filePath string) error {
+	var overwriteErr error
 	for i := 0; i < 3; i++ {
 		if err := overwriteFile(filePath); err != nil {
-			return trace.Wrap(err)
+			if overwriteErr == nil {
+				overwriteErr = err
+			}
 		}
 	}
-	return trace.ConvertSystemError(os.Remove(filePath))
+	// regardless of above errors we want to attempt a removal of the file
+	removeErr := os.Remove(filePath)
+	if overwriteErr == nil {
+		return trace.ConvertSystemError(removeErr)
+	} else if removeErr == nil {
+		return trace.ConvertSystemError(overwriteErr)
+	} else {
+		return trace.Errorf("multiple errors on removal: %v - %v", overwriteErr, removeErr)
```

Comment:
Use `trace.NewAggregate` here instead?

Can get rid of the final else here too.

Reply:
Updated, thank you!